### PR TITLE
Update to CloudEvents SDK 2.0.0-beta.3

### PR DESCRIPTION
**Pull request by jskeet**
Definitely best to review one commit at a time. Note that I wouldn't expect it to all build until we reach the final commit.

### CODE REVIEW

File: src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+                {
+                    throw new InvalidOperationException("Expected Data property to have an object value");
+                }
+                string json = dataElement.GetRawText();
+                cloudEvent.Data = s_jsonParser.Parse<T>(json);
+            }
+
+            protected override void DecodeStructuredModeDataBase64Property(JsonElement dataBase64Element, CloudEvent cloudEvent)
+            {
+                throw new InvalidOperationException("Expected Data property to be set");
```
Comment (amanda-tarafa) [inline src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs:97]: Could you accept here the base64 representation of the binary encoded proto? And maybe emit that as well one method up depending on the CloudEvent.DataContentType?
Would that be useful? I'm not sure.

**Review by amanda-tarafa — COMMENTED** [review]
I've left a few comments here, and I have a few comments/questions about CloudEventFormatter that I'm leaving here as well:

- On CloudEventFormatter, all the `Decode...Message...` methods receive a parameter named `data`, which was confusing, it would have been better if those were called `message`.
- In general I'm a little bit confused about the content type for the whole message and the content type for Data and how that relates to Structured vs Binary. I'm wondering if you considered having two sets of converters, one for the event itself and one for the data so one could more easly mix and match?

File: src/Google.Events.Protobuf.Tests/batch.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+﻿[
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Events.Protobuf.Tests/batch.json:1]: It's a single message (e.g. an HTTP request) but that message contains a batch of two events.

File: src/Google.Events.Protobuf.Tests/binary.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+﻿{
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Events.Protobuf.Tests/binary.json:1]: Possibly. Let's talk about this (and the rest of the terminology here) over VC.

File: src/Google.Events.Protobuf/Cloud/Audit/V1/Data.g.cs [hunk 4]
```diff
@@ -1358,7 +1361,7 @@ public sealed partial class AuditLog : pb::IMessage<AuditLog>
     public const int ServiceDataFieldNumber = 15;
     private global::Google.Protobuf.WellKnownTypes.Struct serviceData_;
     /// <summary>
-    /// Deprecated, use `metadata` field instead.
+    /// Deprecated: Use `metadata` field instead.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Events.Protobuf/Cloud/Audit/V1/Data.g.cs:1364]: The protos aren't marked as obsolete, unfortunately :(

File: src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    /// <summary>
+    /// <see cref="CloudEventFormatter"/> which is able to serialize protobuf messages in JSON.
+    /// (Note that this does not implement the protobuf CloudEvent format, which is separate.)
+    /// </summary>
+    /// <typeparam name="T">The type of message to serialize/deserialize</typeparam>
+    public class ProtobufJsonCloudEventFormatter<T> : CloudEventFormatter where T : IMessage<T>, new()
+    {
+        private static readonly JsonParser s_jsonParser = new JsonParser(JsonParser.Settings.Default.WithIgnoreUnknownFields(true));
+
+        // Note: although we delegate to JsonEventFormatter and *could* just derive from it, that would provide
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs:36]: Clarified to:

```csharp
// Note: although we delegate to a JsonEventFormatter (via StructuredEventFormatter) and *could* just derive from JsonEventFormatter,
// that would provide less flexibility for the future.
```

(JsonEventFormatter is the crucial part here - if we were going with inheritance, we'd derive from that.)

File: src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs (hunk 1 above)
Comment (jskeet) [reply src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs:97]: We *could( potentially do that in the future, yes - although I *suspect* we'd want to just use the protobuf event format (which is a whole other thing) instead.

(I originally optionally used the binary representation in EncodeBinaryModeEventData based on the content type, then decided not to - it's too forward-looking with no real reason to expect that that will happen.)

File: src/Google.Events.Tests/CloudEventFormattersTest.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+            var converter2 = CloudEventFormatters.GetCloudEventFormatter(typeof(EventData));
+            Assert.Same(converter1, converter2);
+        }
+
+        [Fact]
+        public void GetDataConverter_NoCtor() =>
+            Assert.Throws<InvalidOperationException>(() => CloudEventFormatters.GetCloudEventFormatter(typeof(NoCtorData)));
+
+        [Fact]
+        public void GetDataConverter_DoesntImplementInterface() =>
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Events.Tests/CloudEventFormattersTest.cs:38]: Ack - but more importantly, GetDataConverter should be GetCloudEventFormatter in all of these tests :) Fixed.

File: src/Google.Events/CloudEventFormatters.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+                var attribute = targetType.GetCustomAttribute<CloudEventFormatterAttribute>(true);
+                if (attribute is null)
+                {
+                    return null;
+                }
+
+                // It's fine for the converter creation to fail: we'll try it on every attempt,
+                // and always end up with an exception.
+                var converterType = attribute.FormatterType;
+                if (converterType is null)
```
[the remark this answers is not in the document]
Comment (jskeet) [reply src/Google.Events/CloudEventFormatters.cs:54]: Yup: `[CloudEventFormatter(null)]` would be valid, but unhelpful.

**Review by jskeet — COMMENTED** [review]
Thanks for the review. Let's fine some time (maybe tomorrow) to chat about naming.

**Comment by jskeet** [comment]
Now updated to use 2.0.0-beta.3, which simplifies things a lot. PTAL @amanda-tarafa

**Comment by jskeet** [comment]
(Build error is strange - happy to fix that...)

File: src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs (hunk 1 above)
Comment (amanda-tarafa) [reply src/Google.Events.Protobuf/ProtobufJsonCloudEventFormatter.cs:97]: OK, yep, postponing for if/when is needed is totally fine.

File: src/Google.Events.Protobuf.Tests/ProtobufJsonCloudEventFormatterTest.cs [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            AssertJsonBytesEqual(expectedBytes, actualBytes);
+        }
+
+        private static void AssertJsonBytesEqual(byte[] expected, byte[] actual)
+        {
+            JToken expectedObject = Parse(expected);
+            JToken actualObject = Parse(actual);
+            Assert.Equal(expectedObject, actualObject);
+
+            JToken Parse(byte[] bytes)
```
Comment (amanda-tarafa) [inline src/Google.Events.Protobuf.Tests/ProtobufJsonCloudEventFormatterTest.cs:170]: Is this why you took the Newtonsoft dependency right? There's no easy way to do this with System.Json? Just curious mostly.

File: src/Google.Events.Protobuf.Tests/CloudEventFormatterAttributesTest.cs [hunk 9]
```diff
@@ -0,0 +1,42 @@
+﻿// Copyright 2020, Google LLC
```
[the remark this answers is not in the document]
Comment (amanda-tarafa) [reply src/Google.Events.Protobuf.Tests/CloudEventFormatterAttributesTest.cs]: ping ;)

**Review by amanda-tarafa — APPROVED** [review]
LGTM!

File: src/Google.Events.Protobuf.Tests/ProtobufJsonCloudEventFormatterTest.cs (hunk 8 above)
Comment (jskeet) [reply src/Google.Events.Protobuf.Tests/ProtobufJsonCloudEventFormatterTest.cs:170]: Json.NET is definitely easier to use for this sort of thing, yes. It's just a bigger API with more convenience methods etc.

**Comment by jskeet** [comment]
Will fix the copyright year then merge - thanks!

**Pull request merged**